e2e/js_image/js_image_layer.bzl

```diff
@@ -0,0 +1,247 @@
+            file_map.pop(destination)
+        info = PackageSymlinkInfo(
+            target = "/%s" % _runfile_path(ctx, symlink.target_file, runfiles_dir),
+            destination = destination,
+            attributes = {"mode": "0777"},
+        )
+        symlinks.append([info, symlink.target_file.owner])
+
+    for symlink in default.data_runfiles.root_symlinks.to_list():
+        destination = "/".join([runfiles_dir, symlink.path])
+        if not _should_include(destination, ctx.attr.include, ctx.attr.exclude):
+            continue
+        if hasattr(file_map, destination):
+            file_map.pop(destination)
+        info = PackageSymlinkInfo(
+            target = "/%s" % _runfile_path(ctx, symlink.target_file, runfiles_dir),
+            destination = destination,
+            attributes = {"mode": "0777"},
+        )
+        symlinks.append([info, symlink.target_file.owner])
+
+    return [
+        PackageFilegroupInfo(
+            pkg_dirs = [],
+            pkg_files = [
+                [PackageFilesInfo(
+                    dest_src_map = file_map,
+                    attributes = {},
+                ), ctx.label],
+            ],
+            pkg_symlinks = symlinks,
+        ),
+        DefaultInfo(files = files),
+    ]
+
+runfiles = rule(
+    implementation = _runfiles_impl,
+    attrs = {
+        "binary": attr.label(mandatory = True),
+        "root": attr.string(),
+        "include": attr.string(),
+        "exclude": attr.string(),
+    },
+)
+
+BUILD_TAR = """
+import os
+import sys
+import json
+import subprocess
+from rules_python.python.runfiles import runfiles
+
+r = runfiles.Create()
+build_tar_path = r.Rlocation("rules_pkg/pkg/private/tar/build_tar")
+
+manifest_path = next((argv.replace("--manifest=", "") for argv in sys.argv if argv.startswith("--manifest=")), None)
+
+with open(manifest_path, 'r') as manifest_fp:
+    manifest = json.load(manifest_fp)
+
+def strip_execroot(p):
+    parts = p.split(os.sep)
+    i = parts.index("execroot") 
+    return os.sep.join(parts[i+2:]) 
+
+def get_runfiles_path(p):
+    for entry in manifest:
+        if entry[2] == p:
+            return entry[1]
+    raise Exception("could not find a corresponding file for %s within manifest.")
+
+
+for entry in manifest:
+    p = entry[2]
+    if "node_modules" in p and os.path.islink(p):
+        link_to = os.path.realpath(p)
+        link_to_execroot_stripped = strip_execroot(link_to)
+        overwritten_to = get_runfiles_path(link_to_execroot_stripped)
+        # fix it!
+        entry[0] = 1 # make it a symlink
+        entry[2] = "/%s" % overwritten_to
+        entry[3] = "0777"
+
+os.remove(manifest_path)
+with open(manifest_path, "w") as manifest_w:
+    manifest_w.write(json.dumps(manifest))
+
+r = subprocess.run([build_tar_path] + sys.argv[1:])
+sys.exit(r.returncode)
+"""
+
+def js_image_layer(name, binary, root = None, **kwargs):
+    """Creates a tar file containing runfiles from the binary
```

Comment:
Docs out of date here since this now creates two tar files.

Probably good to also mention that the entry point will be `/{root of js_image_layer}/{package_name()}/{name of js_binary}.sh` and add a brief snippet of usage into the docstring.

Are you thinking this will serve as an example for now and in a follow-up could be promoted to the public API?

Reply:
> Are you thinking this will serve as an example for now and in a follow-up could be promoted to the public API?

Exactly. We should promote this to public API before 1.0 with a follow up PR.